File: loomoroutines-dispatcher/src/main/java/dev/reformator/loomoroutines/dispatcher/utils/DispatcherUtils.java
package dev.reformator.loomoroutines.dispatcher.utils;

import dev.reformator.loomoroutines.dispatcher.Dispatcher;
import dev.reformator.loomoroutines.dispatcher.Promise;
import dev.reformator.loomoroutines.utils.CoroutineUtils;
import dev.reformator.loomoroutines.common.internal.utils.Mutable;
import dev.reformator.loomoroutines.dispatcher.internal.DispatcherContext;
import dev.reformator.loomoroutines.dispatcher.internal.DispatcherContextImpl;
import dev.reformator.loomoroutines.dispatcher.internal.utils.InternalDispatcherUtils;
import org.jetbrains.annotations.NotNull;

import java.time.Duration;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class DispatcherUtils {
    private DispatcherUtils() { }

    private static final Supplier<Void> nullGetter = () -> null;

    public static void delay(@NotNull Duration duration) {
        Objects.requireNonNull(duration);
        var coroutine = getDispatcherCoroutine();
        coroutine.getCoroutineContext().setDelayLastEvent(duration);
        coroutine.suspend();
    }

    public static void delay(long millis) {
        delay(Duration.ofMillis(millis));
    }

    public static void executeInDispatcher(@NotNull Dispatcher newDispatcher, @NotNull Runnable action) {
        Objects.requireNonNull(newDispatcher);
        Objects.requireNonNull(action);
        var coroutine = getDispatcherCoroutine();
        var oldDispatcher = Objects.requireNonNull(coroutine.getCoroutineContext().getDispatcher());
        if (oldDispatcher == newDispatcher) {
            action.run();
        } else {
            coroutine.getCoroutineContext().setSwitchLastEvent(newDispatcher);
            coroutine.suspend();
            try {
                action.run();
            } finally {
                var newCoroutine = getDispatcherCoroutine();
                newCoroutine.getCoroutineContext().setSwitchLastEvent(oldDispatcher);
                newCoroutine.suspend();
            }
        }
    }

    public static <T> @NotNull T executeInDispatcher(
            @NotNull Dispatcher newDispatcher,
            @NotNull Supplier<? extends T> action
    ) {
        Objects.requireNonNull(action);
        var result = new Mutable<T>();
        executeInDispatcher(newDispatcher, () -> result.set(action.get()));
        return result.get();
    }

    public static void await(@NotNull Consumer<? super Runnable> subscribeFunc) {
        Objects.requireNonNull(subscribeFunc);
        var coroutine = getDispatcherCoroutine();
        coroutine.getCoroutineContext().setAwaitLastEvent(subscribeFunc);
        coroutine.suspend();
    }

    public static <T> @NotNull Promise<T> dispatch(@NotNull Dispatcher dispatcher, @NotNull Supplier<T> action) {
        var context = new DispatcherContextImpl<T>();
        var result = new Mutable<T>();
        var coroutine = CoroutineUtils.createCoroutine(
                context,
                () -> result.set(action.get())
        );
        InternalDispatcherUtils.dispatchCoroutine(dispatcher, coroutine, result::get);
        return context.getPromise();
    }

    public static @NotNull Promise<Void> dispatch(@NotNull Dispatcher dispatcher, @NotNull Runnable action) {
        var context = new DispatcherContextImpl<Void>();
        var coroutine = CoroutineUtils.createCoroutine(context, action);
        InternalDispatcherUtils.dispatchCoroutine(dispatcher, coroutine, nullGetter);
        return context.getPromise();
    }

    @SuppressWarnings("unchecked")
    private static RunningCoroutine<? extends DispatcherContext<?>> getDispatcherCoroutine() {
        var coroutine = CoroutineUtils.getRunningCoroutineByContextType(DispatcherContext.class);
        if (coroutine == null) {
            throw new IllegalStateException("Not in dispatcher context");
        }
        return (RunningCoroutine<? extends DispatcherContext<?>>) coroutine;
    }
}

File: loomoroutines-dispatcher/src/main/kotlin/dispatcher-utils.kt
@file:JvmName("DispatcherUtils")

package dev.reformator.loomoroutines.dispatcher

import dev.reformator.loomoroutines.common.createCoroutine
import dev.reformator.loomoroutines.common.getRunningCoroutineContext
import dev.reformator.loomoroutines.common.internal.Consumer
import dev.reformator.loomoroutines.common.internal.Supplier
import dev.reformator.loomoroutines.common.internal.invoke
import dev.reformator.loomoroutines.common.internal.kotlinstdlibstub.Ref
import dev.reformator.loomoroutines.common.trySuspendCoroutine
import dev.reformator.loomoroutines.dispatcher.internal.DispatcherContext
import dev.reformator.loomoroutines.dispatcher.internal.DispatcherContextImpl
import dev.reformator.loomoroutines.dispatcher.internal.ScheduledExecutorServiceDispatcher
import dev.reformator.loomoroutines.dispatcher.internal.dispatch
import java.time.Duration
import java.util.concurrent.ScheduledExecutorService

/**
 * Optional annotation to warn that the method must be called only inside a dispatcher coroutine.
 */
@Target(AnnotationTarget.FUNCTION)
annotation class CallOnlyInDispatcher

/**
 * Is current thread running in a dispatcher coroutine.
 */
val isInDispatcher: Boolean
    get() = getRunningCoroutineContext<DispatcherContext<*>>() != null

/**
 * Notifier about completing of an awaiting.
 * @see await
 */
fun interface Notifier {
    /**
     * Notify the dispatcher coroutine that the awaiting has completed.
     * Must be called only once for each [Notifier].
     * @see await
     */
    operator fun invoke()
}

/**
 * Suspend a dispatcher coroutine until a [Notifier] will be invoked.
 * The method suspend a dispatcher coroutine and call [callback] with a [Notifier] which must be invoked to continue the coroutine.
 */
@CallOnlyInDispatcher
fun await(callback: Consumer<Notifier>) {
    sendDispatcherEvent {
        setAwaitLastEvent(callback)
    }
}

/**
 * Suspend a dispatcher coroutine and continue it after a [delay][duration].
 */
@CallOnlyInDispatcher
fun delay(duration: Duration) {
    sendDispatcherEvent {
        setDelayLastEvent(duration)
    }
}

/**
 * Execute an [action] possibly switching a [Dispatcher].
 * @param dispatcher a dispatcher in which [action] will be executed
 * @return [action]'s result
 */
@CallOnlyInDispatcher
fun <T> doIn(dispatcher: Dispatcher, action: Supplier<T>): T {
    val context = getMandatoryRunningCoroutineDispatcherContext()
    val oldDispatcher = context.dispatcher!!
    if (oldDispatcher === dispatcher) {
        return action()
    } else {
        sendDispatcherEvent {
            setSwitchLastEvent(dispatcher)
        }
        val result = try {
            action()
        } finally {
            sendDispatcherEvent {
                setSwitchLastEvent(oldDispatcher)
            }
        }
        return result
    }
}

/**
 * Create a dispatcher coroutine that will run in [this] and execute [body].
 */
fun <T> Dispatcher.dispatch(body: Supplier<T>): Promise<T> {
    val context = DispatcherContextImpl<T>()
    val result = Ref.ObjectRef<T>()
    val coroutine = createCoroutine(context) { result.element = body() }
    dispatch(coroutine, result)
    return context.promise
}

/**
 * Create a [CloseableDispatcher] that executes actions in [this].
 */
fun ScheduledExecutorService.toDispatcher(): CloseableDispatcher =
    ScheduledExecutorServiceDispatcher(this)

@CallOnlyInDispatcher
private inline fun sendDispatcherEvent(crossinline generateEvent: DispatcherContext<*>.() -> Unit) {
    val suspended = trySuspendCoroutine<DispatcherContext<*>> {
        it.generateEvent()
        true
    }
    if (!suspended) {
        throwNotInDispatcher()
    }
}

@CallOnlyInDispatcher
private fun getMandatoryRunningCoroutineDispatcherContext(): DispatcherContext<*> =
    getRunningCoroutineContext<DispatcherContext<*>>() ?: throwNotInDispatcher()

@CallOnlyInDispatcher
private fun throwNotInDispatcher(): Nothing =
    error("Method must be called in a dispatcher coroutine.")
